src/gdbstub_impl/ext/host_io.rs

```diff
@@ -0,0 +1,69 @@
+use super::prelude::*;
+use crate::arch::Arch;
+use crate::protocol::commands::ext::HostIo;
+use crate::target::ext::host_io::PreadOutput;
+
+impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+    pub(crate) fn handle_host_io(
+        &mut self,
+        res: &mut ResponseWriter<C>,
+        target: &mut T,
+        command: HostIo,
+    ) -> Result<HandlerStatus, Error<T::Error, C::Error>> {
+        let ops = match target.host_io() {
+            Some(ops) => ops,
+            None => return Ok(HandlerStatus::Handled),
+        };
+
+        crate::__dead_code_marker!("host_io", "impl");
+
+        let handler_status = match command {
+            HostIo::vFileOpen(cmd) => {
+                let ret = ops.open(cmd.filename, cmd.flags, cmd.mode);
+                res.write_str("F")?;
+                res.write_num(ret)?;
+                HandlerStatus::Handled
+            }
+            HostIo::vFileClose(cmd) => {
+                let ret = ops.close(cmd.fd);
+                res.write_str("F")?;
+                res.write_num(ret)?;
+                HandlerStatus::Handled
+            }
+            HostIo::vFilePread(cmd) => {
+                let count = <T::Arch as Arch>::Usize::from_be_bytes(cmd.count)
+                    .ok_or(Error::TargetMismatch)?;
+                let offset = <T::Arch as Arch>::Usize::from_be_bytes(cmd.offset)
+                    .ok_or(Error::TargetMismatch)?;
+                let mut err: Result<_, Error<T::Error, C::Error>> = Ok(());
+                let mut callback = |data: &[u8]| {
+                    let e = (|| {
+                        res.write_str("F")?;
+                        res.write_num(data.len())?;
```

Comment:
ahh, shoot, I didn't realize that you'd need the data's length at the start of the packet...

In that case, we need to tweak the API of `PreadOutput` to make it impossible to call twice. i.e: change the signature of `PreadOutput::write` to take `self` instead of `&mut self`. This also means that the API of the `pread` handler will need to be switched from `output: &mut PreadOutput` to just `output: PreadOutput`.

* * * 

In addition, we'd want some way to enforce that the target has in-fact called `PreadOutput::write`. if the target is misbehaving and not calling write, I don't think returning a blank response is the correct approach.

I suspect there might be a fun little phantom-data, "abuse the type-system + lifetimes" way guarantee this property, but I don't have time to sketch out a solution right now. Maybe later tonight...

Reply:
Hope for details

Reply:
Ahh, yes, sorry about that. I did in-fact have some time to noodle around with a type-safe solution. Here's what I came up with:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6f6ce57f616f90a233afc2beeb25cac8

Essentially, we'd change the return type of `pread` to return a `HostIoOutputToken<'a>` Zero Sized Type, where the `'a` lifetime is tied to that specific invocation of `pread`. 

The only way to acquire a `HostIoOutputToken` is by calling `HostIoOutput::write`, which in-turn consumes `HostIoOutput`, and doesn't allow any additional `HostIoOutputToken`s to be constructed. Plus, `HostIoOutputToken` is no Send/Sync, nor Copy/Clone, so it should be impossible to "smuggle" it out of the `pread` context for later use (assuming the implementation was actively trying to shoot itself in the foot).

This entire dance will not incur any runtime overhead, since `HostIoOutputToken` is a ZST.

Pretty neat, eh?

Reply:
Hope I integrate it right 😅

Reply:
Yeah, you did! Awesome!